### PR TITLE
feat: add convert_labelbox utility for Labelbox NDJSON to YOLO convert

### PR DESCRIPTION
**Pull request by mdayku**
feat: add convert_labelbox utility for Labelbox NDJSON → YOLO conversion

## Summary

This PR adds a `convert_labelbox` utility to the Ultralytics data converters, providing a first-class way to convert modern Labelbox NDJSON exports into YOLO **detection** datasets (images, labels, and a dataset YAML).

The converter is intentionally small and focused on bounding-box annotations plus a simple YAML-based class mapping, rather than a full Labelbox integration.

## Motivation

- Ultralytics already provides `convert_coco` and related converters for popular formats.
- Labelbox is widely used for annotation, but conversion currently relies on:
  - Legacy scripts (e.g. JSON2YOLO), or
  - Custom, one-off converters written by users.
- Those older scripts are not integrated into the modern `ultralytics.data.converter` API.
- This PR brings Labelbox support into the main library in a maintainable, explicitly-configured way.

## What This PR Does

- Adds `ultralytics/data/labelbox.py` with:
  - `load_class_map(path)`: map class name → class ID based on a YAML file.
  - `load_labelbox_mapping(path)`: map Labelbox label strings → class ID.
  - `parse_labelbox_ndjson(path)`: yield rows from a Labelbox NDJSON export.
  - `convert_labelbox(labels_path, images_dir, save_dir, class_map)`: main conversion function for **bounding boxes only**.
- Re-exports `convert_labelbox` from `ultralytics.data.converter`.
- Adds tests in `tests/test_labelbox_convert.py`:
  - `test_load_class_map_and_mapping` checks that:
    - Class IDs are assigned in the expected order.
    - Labelbox labels map correctly to class IDs.
  - `test_convert_labelbox_creates_yolo_dataset`:
    - Builds a tiny NDJSON + dummy image.
    - Runs `convert_labelbox(...)`.
    - Asserts that `data.yaml`, one label file, and the copied image are created with the expected normalized bbox.
- Adds a usage page: `docs/en/usage/convert_labelbox.md`.

## Usage & Workflow (not plug-and-play)

This utility is a **conversion helper**, not a full Labelbox integration.

**Prerequisites:**

- You have already exported your project from Labelbox as **NDJSON**.
- You have downloaded the corresponding images referenced by `externalId` into a local `images_dir`.
- You have created a simple **class map YAML** (e.g. `class_map.yaml`) that maps Labelbox label strings → YOLO class IDs.

**Example:**

from ultralytics.data.converter import convert_labelbox

convert_labelbox(
    labels_path="export.ndjson",
    images_dir="images/",
    save_dir="datasets/myproject",
    class_map="class_map.yaml",
)This will create a YOLO-style folder structure under `save_dir` and a `data.yaml` using the class names and IDs from `class_map.yaml`.

## Non-Goals (v1)

This PR intentionally does **not**:

- Call the Labelbox API or download images for you (no API keys, no network calls).
- Handle segmentation / polygon annotations (detection bounding boxes only).
- Implement multi-map conversion or advanced schema logic.
- Automatically adapt to arbitrary NDJSON schemas:
  - It assumes a Labelbox-style NDJSON with keys like `externalId`, `imageWidth`, `imageHeight`, and `objects[*].bbox`.

These constraints keep the implementation small, predictable, and low-risk.

## Testing

- Local tests:

python -m pytest tests/test_labelbox_convert.py- Both tests pass on my local environment (Windows, Python 3.11).

## Related Issues

This PR targets `ultralytics/ultralytics`, but it is motivated by recurring
Labelbox → YOLO conversion friction across the Ultralytics ecosystem:

- Related: https://github.com/ultralytics/yolov5/issues/11773  
  “Using LabelBox” – user notes that the legacy `labelbox_json2yolo.py`
  script no longer works after Labelbox changed its export format, and that
  the suggested `export_to_coco.py` + `general_json2yolo.py` pipeline also
  fails for their use case.

- Related: https://github.com/ultralytics/JSON2YOLO/issues/102  
  “Labelbox2YOLO Json.load issue” – open bug when trying to convert
  Labelbox exports with the helper script in the JSON2YOLO repo.

- Related: https://github.com/ultralytics/JSON2YOLO/issues/39  
  “Converting Labelbox segmentation json to coco” – user asking how to make
  Labelbox “Mask” exports work with YOLO; highlights that Labelbox formats
  are non-trivial to handle and currently require custom code.

- Related: https://github.com/ultralytics/JSON2YOLO/issues/37  
  “Solutions to 'make_dirs' problem in labelbox_json2yolo.py” – users
  running into errors with the older standalone Labelbox → YOLO script.

- Related: https://github.com/ultralytics/JSON2YOLO/issues/53  
  “There are absolutely no explanations on how to run the code or what
  parameters one should use” – general documentation confusion around how
  to use the JSON2YOLO tooling, which affects Labelbox users as well.

While this PR does not directly close those issues (they live in other
repositories), `convert_labelbox` provides a supported, documented path
for modern Labelbox NDJSON → YOLO **detection** datasets within the main
Ultralytics library, which should reduce the need for ad-hoc scripts and
workarounds described above.

## CLA

I have read the CLA Document and I sign the CLA.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a new `convert_labelbox` utility, tests, and documentation to convert Labelbox NDJSON exports into YOLO-format detection datasets.

### 📊 Key Changes
- ➕ Introduces `ultralytics.data.labelbox` with `convert_labelbox`, `load_class_map`, `load_labelbox_mapping`, and `parse_labelbox_ndjson` utilities.
- 🧪 Adds `tests/test_labelbox_convert.py` to validate class-map loading, Labelbox-to-YOLO mapping, and full NDJSON-to-YOLO dataset conversion (including bbox normalization and image copying).
- 📚 Adds `docs/en/usage/convert_labelbox.md` describing prerequisites, Python usage, and a proposed `yolo convert labelbox` CLI workflow.

### 🎯 Purpose & Impact
- 🧩 Simplifies migrating Labelbox NDJSON detection annotations into YOLO training datasets with a single helper.
- 🛠 Provides a clear, YAML-based mapping from Labelbox class names to YOLO class IDs, improving reproducibility and transparency.
- 🚀 Lowers the friction for Labelbox users to train YOLO models by automating dataset structure, label file creation, and dataset YAML generation.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mdayku, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated review helper, and an Ultralytics engineer will follow up to assist with the next steps.

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/data/labelbox.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            y_center = (top + bh / 2.0) / float(height)
+            w_norm = bw / float(width)
+            h_norm = bh / float(height)
+
+            line = f"{cls_id} {x_center:.6f} {y_center:.6f} {w_norm:.6f} {h_norm:.6f}"
+            label_lines.append(line)
+            n_boxes += 1
+
+        if label_lines:
+            n_images += 1
```
Comment (UltralyticsAssistant) [inline ultralytics/data/labelbox.py]: ⚠️ **HIGH**: Label files are always written as `labels/train/<stem>.txt`, so nested `externalId` values like `floor1/cam0.jpg` will collide (`cam0.txt`) and no matching subdirectories are created. This breaks conversions that rely on hierarchical image names and can silently overwrite annotations. Preserve the relative path (including folders) and ensure its parent directory exists before writing.

File: docs/en/usage/convert_labelbox.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+It **does not**:
+
+- Call the Labelbox API or download images for you.
+- Handle segmentation/polygon annotations in v1 (bounding boxes only).
+- Automatically adapt to arbitrary NDJSON schemas; it assumes a Labelbox-style schema with fields like `externalId`, `imageWidth`, `imageHeight`, and `objects[*].bbox`.
+
+## Python
+
+```python
```
Comment (UltralyticsAssistant) [inline docs/en/usage/convert_labelbox.md:24]: 💡 **MEDIUM**: The example import points to `ultralytics.data.converter`, but the new helper is defined in `ultralytics.data.labelbox`. Using the wrong module path will raise `ImportError` for users following the docs. Please update the sample code to import from the newly added module.

**Suggested change:**
```suggestion
from ultralytics.data.labelbox import convert_labelbox
```

File: ultralytics/data/labelbox.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+            x_center = (left + bw / 2.0) / float(width)
+            y_center = (top + bh / 2.0) / float(height)
+            w_norm = bw / float(width)
+            h_norm = bh / float(height)
+
+            line = f"{cls_id} {x_center:.6f} {y_center:.6f} {w_norm:.6f} {h_norm:.6f}"
+            label_lines.append(line)
+            n_boxes += 1
+
```
Comment (UltralyticsAssistant) [inline ultralytics/data/labelbox.py:162]: 💡 **MEDIUM**: When an image contains no kept objects (unknown class, zero-size boxes, etc.) the code copies the image but skips writing an empty label file. Ultralytics dataloaders expect a label file for every image; missing files trigger warnings and unnecessary filesystem lookups. Consider touching an empty `.txt` (or skipping the image copy) to keep the dataset consistent.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Update the documentation import to reference the new module, and adjust `convert_labelbox` to preserve subdirectories in label paths and to emit empty label files (or skip copying) when no annotations are retained.

💬 Posted 3 inline comments

**Comment by mdayku** [comment]
I have read the CLA Document and I sign the CLA

File: ultralytics/data/labelbox.py (hunk 1 above)
Comment (mdayku) [reply ultralytics/data/labelbox.py]: > Fixed. Label files now preserve the relative path from externalId (e.g., floor1/cam0.jpg → floor1/cam0.txt) and create parent directories as needed.

File: ultralytics/data/labelbox.py (hunk 3 above)
Comment (mdayku) [reply ultralytics/data/labelbox.py:162]: > Fixed. An empty .txt file is now written for images with no kept annotations, matching YOLO dataloader expectations.

File: docs/en/usage/convert_labelbox.md (hunk 2 above)
Comment (mdayku) [reply docs/en/usage/convert_labelbox.md:24]: > Fixed. Updated to from ultralytics.data.labelbox import convert_labelbox.

**Review by TylerPohn — COMMENTED** [review]
lgtm
